### DISCUSSION TITLE
Compatibility with data APIs in react router 6.4

### DISCUSSION THREAD

**Discussion by codyzu** [urql-graphql/urql#2699] [Q&A] [answered]
React router is now promoting coupling of data and routing: https://remix.run/blog/react-router-v6.4

Are there plans to make urql more compatible with the [new data APIs](https://reactrouter.com/en/main/start/overview#data-loading) that landed in react router 6.4?

Building an app with react router and urql now gives 2 different ways to handle things like waiting for data to be loaded. The new data APIs in react router look interesting because they allow [pending navigation](https://reactrouter.com/en/main/start/overview#pending-navigation-ui), [integrate with react Suspense](https://reactrouter.com/en/main/start/overview#skeleton-ui-with-suspense), and do some [fancy form handling](https://reactrouter.com/en/main/start/overview#data-mutations).

AFAIK, the primary useQuery like urql hooks aren't compatible out of the box because the routes are now defined outside of a component render (no hooks here).

I'm a big promoter of urql (it will be a big part of my talk at nodeconf.eu next month) and I'm wondering if there are plans to make react router and urql more compatible and what your advice would be for devs picking up urql today.

Thanks for all of the amazing work!

**Comment by JoviDeCroock** [comment, marked as answer]
You can use the result of `createClient` to load data in those routes, it will use the same exchanges so data aggregated in caches will benefit from that. This means that if you do `loader={() => { await client.query().toPromise() }}` that same data will be available to a `useQuery` invocation given the same query (or in a normalized cache a subset of it).

Is there any specific issue you are facing with this integration? It's hard to improve on user-provided functions as we can't automate the process with specific libraries - but the core urql methods are compatible with anything and I personally don't think it's that much more work 😅 

Good luck on your talk, would love to hear more about it, feel free to tweet it at me if there's a recording or where we can see it live!

**Reply by codyzu** [reply]
Thanks for this response. I'm going to give it a try and then I'll accept the answer.

I'll be sure to tweet about the talk!!!

**Reply by JoviDeCroock** [reply]
@codyzu any update on this? 😅 did not have the time yet to test this myself but I would assume it just works unless there's some magic going on

**Reply by villesau** [reply]
Does urql support abort signals? https://reactrouter.com/en/main/route/loader

I also think that most of the type safety is lost if using the route loaders, and there's probably no easy way around it due to the way react router is built :/ Data is fetched in loader but used via some hook so there is no proper coupling that would carry over the types.

**Reply by JoviDeCroock** [reply]
https://github.com/FormidableLabs/urql/discussions/2749#discussioncomment-3901348 there's an answer here, not directly but you could wrap fetch
